src/client.rs

```diff
@@ -0,0 +1,260 @@
+use crate::errors::ConnectorError;
+use reqwest::{Client, Response, StatusCode};
+use url::{ParseError, Url};
+
+const DEFAULT_API_BASE_URL: &str = "https://api.meteomatics.com";
+
+#[derive(Clone, Debug)]
+pub struct APIClient {
+    http_client: Client,
+    username: String,
+    password: String,
+}
+
+impl APIClient {
+    pub fn new(username: String, password: String, timeout_seconds: u64) -> Self {
+        // safe to use unwrap, since we want to panic if the client builder fails.
+        let http_client = Client::builder()
+            .timeout(std::time::Duration::from_secs(timeout_seconds))
+            .build()
+            .unwrap();
+
+        Self {
+            http_client,
+            username,
+            password,
+        }
+    }
+
+    pub async fn query_time_series(
+        &self,
+        start_date: &chrono::DateTime<chrono::Utc>,
+        end_date: &chrono::DateTime<chrono::Utc>,
+        interval: &chrono::Duration,
+        parameters: &Vec<String>,
+        coordinates: &Vec<Vec<f64>>,
+        optionals: &Option<Vec<String>>,
+    ) -> Result<polars::frame::DataFrame, ConnectorError> {
+
+        // Create the coordinates
+        let coords_str = coords_to_str(&coordinates).await;
+
+        // Create the query specifications (time, location, etc.)
+        let query_specs = build_query_specs(
+            &start_date, &end_date, &interval, &parameters, &coords_str, &optionals
+        ).await;
+
+        // Get the query result
+        let result = self.do_http_get(&query_specs).await;
+
+        // Match the result
+        // TODO: Check this match statement, when is ApiError produced?
+        match result {
+            Ok(response) => match response.status() {
+                StatusCode::OK => {
+                    let dataframe: polars::frame::DataFrame = parse_response_to_dataframe(response)
+                        .await?;
+        
+                    Ok(dataframe)
+                }
+                status => Err(ConnectorError::HttpError(
+                    status.to_string(),
+                    response.text().await.unwrap(),
+                    status,
+                )),
+            },
+            Err(connector_error) => Err(ConnectorError::ApiError {
+                source: connector_error,
+            }),
+        }
+    }
+    
+    async fn do_http_get(&self, url_fragment: &str) -> Result<Response, reqwest::Error> {
+        let full_url = build_url(url_fragment)
+            .await
+            .expect("URL fragment must be valid");
+
+        println!(">>>>>>>>>> full_url: {}", full_url);
+
+        self.http_client
+            .get(full_url)
+            .basic_auth(&self.username, Some(String::from(&self.password)))
```

Comment:
&self.password is already a string, String::from is an excess I would say, why not just
Some(&self.password) ?
```suggestion
            .basic_auth(&self.username, Some(&self.password))
```